### PR TITLE
[EFM] minor polishing

### PR DESCRIPTION
**Pull request by AlexHentschel**
for our `EpochLookup` we have introduced the auxiliary component `epochRangeCache`. While the implementation is highly optimized (using an array of value types as opposed to pointers for improved data locality), it leaked implementation details (specifically the concept of zero values representing the "undefined" value). The documentation was hardly clarifying this special meaning of the zero value, and in [some areas was still (incorrectly) talking about nil values](https://github.com/onflow/flow-go/blob/2aeebea1359c9089633de5608fad1405c1adb1c3/module/epochs/epoch_lookup.go#L37-L38). 

This PR:
* cleans up the documentation
* increases encapsulation so that the `EpochLookup` does not need to concern itself with the implementation detail how  `epochRangeCache` represents the undefined entry. For this purpose:
  - I replaced the method [`epochRangeCache.combinedRange`](https://github.com/onflow/flow-go/blob/2aeebea1359c9089633de5608fad1405c1adb1c3/module/epochs/epoch_lookup.go#L79-L94) with `epochRangeCache.cachedEpochs`
  - consolidated and optimized the logic [`EpochLookup.EpochForView`](https://github.com/onflow/flow-go/blob/2aeebea1359c9089633de5608fad1405c1adb1c3/module/epochs/epoch_lookup.go#L238-L269)

### CODE REVIEW

File: module/epochs/epoch_lookup.go [hunk 1]
```diff
@@ -58,13 +69,13 @@ func (cache *epochRangeCache) extendLatestEpoch(epochCounter uint64, extension f
 	}
 
 	// sanity check: `extension.FinalView` should be greater than final view of latest epoch
-	if cache[2].finalView > extension.FinalView {
+	if latestEpoch.finalView > extension.FinalView {
 		return fmt.Errorf(invalidExtensionFinalView, cache[2].finalView, extension.FinalView)
```
Comment (durkmurder) [inline module/epochs/epoch_lookup.go:73]:
```suggestion
		return fmt.Errorf(invalidExtensionFinalView, latestEpoch.finalView, extension.FinalView)
```

File: module/epochs/epoch_lookup.go [hunk 2]
```diff
@@ -58,13 +69,13 @@ func (cache *epochRangeCache) extendLatestEpoch(epochCounter uint64, extension f
...
 	}
 
 	// sanity check: epoch extension should have the same epoch counter as the latest epoch
-	if cache[2].counter != epochCounter {
-		return fmt.Errorf(mismatchEpochCounter, cache[2].counter, epochCounter)
+	if latestEpoch.epochCounter != epochCounter {
+		return fmt.Errorf(mismatchEpochCounter, cache[2].epochCounter, epochCounter)
```
Comment (durkmurder) [inline module/epochs/epoch_lookup.go:78]:
```suggestion
		return fmt.Errorf(mismatchEpochCounter, latestEpoch.epochCounter, epochCounter)
```

**Review by durkmurder — APPROVED** [review]
Looks good, it's a bit complex without using pointers but I guess that is the price that we pay for better data locality.

File: module/epochs/epoch_lookup.go [hunk 3]
```diff
@@ -237,36 +248,39 @@ func (lookup *EpochLookup) cacheEpoch(epoch protocol.Epoch) error {
...
-		if !epoch.exists() {
-			continue
-		}
-		if epoch.firstView <= view && view <= epoch.finalView {
-			return epoch.counter, nil
+	// -------[----|-*--|----]-------
+	for i := l - 1; i >= 0; i-- {
+		if cachedEpochs[i].firstView <= view {
+			// we know: view ≤ cachedEpochs[i].finalView, because otherwise we would have returned via a prior
+			// value check triggering; or epochRangeCache has a bug, which we avoid with thorough testing.
```
Comment (jordanschalm) [inline module/epochs/epoch_lookup.go:277]:
```suggestion
		if cachedEpochs[i].firstView <= view && view <= cachedEpochs[i].finalView {
```

I agree this is correct but I feel like we can afford the extra nanosecond it takes to make this conditional auditable for correctness without reading the whole function. Let's _use_ that data locality! ;)

**Review by jordanschalm — APPROVED** [review]

File: module/epochs/epoch_lookup.go (hunk 3 above)
Comment (AlexHentschel) [reply module/epochs/epoch_lookup.go:277]: 💯

**Pull request merged**